### PR TITLE
Real-to-real transforms (version 3)

### PR DESCRIPTION
**Pull request by grlee77**
I have rebased the nice work by @drwells from #95 on current master, so the first 20 or so commits here are the ones from that PR with modifications made only as necessary to resolve conflicts.

That initial rebase went fairly quickly, but then it took an hour or two to track down a few mistakes I made  (fixed in 527c4e3 and d7ee369).

I updated the `threads` and `planner_effort` handling to be consistent with pyFFTW 0.11+ and added the n-dimensional variants of the DCT and DST introduced in SciPy 1.0. Currently these are implemented by separable application of 1D transforms rather than trying to implement an n-dimensional R2R builder.

I have copied over the TODO list from a comment in #95 below for quick reference:

- [ ] Tests for builder functions
- [ ] Tests for failure modes
- [x] Improve SciPy interface's handling of `None` with orthogonalization
- [ ] Implement padding and truncation for the `dct` and `dst` functions
- [x] Clean up scipy tests
- [x] Clean up real to real tests (use existing infrastructure)
- [ ] Improve `_Xfftn`
- [x] Update documentation to describe real to real features 

New items I added
- [x] flexible defaults for threads and planner_effort in the builders and interfaces
- [x] dctn, idctn, dstn, dctn added to interfaces.scipy_fftpack
- [x] test axis argument of dct, etc. in the scipy interfaces tests
- [x] add tests for n-dimensional R2R transforms in the scipy interfaces
- [ ] n-dimensional builders for dctn, dstn?

If anyone want to volunteer to work on any of the remaining items, please do so.

### CODE REVIEW

**Comment by rabernat** [comment]
Hi @grlee77 - thanks for picking up this issue. It's beyond my expertise to contribute directly, but I just wanted to state that this issue is very important to the pyqg project. We have been hoping for it for over three years! (pyqg/pyqg#132)

Cython expertise is very rare in my world. I wish I could offer to help more directly. But at least let me say, THANK YOU! 👏  Your efforts are sincerely appreciated.

**Comment by grlee77** [comment]
Thanks for the encouragement @rabernat! I had looked at your Pangeo project a little in the past and it looked nice, but have not tried to use it (I work in medical imaging rather than earth sciences).

It is possible that this will get some attention as part of an ongoing Scipy GSOC 2019 project later this summer. That project primarily involves creating a system to allow flexible choice of FFT backends (pyFFTW, mkl-fft, etc) from SciPy itself, but may involve some contributions here to fill gaps in the scipy API (i.e. real-to-real transform support). Step 1 of the project (adding a new `scipy.fft` module) has been merged (scipy/scipy#10238), and prototypes of the backend system are in progress. 

I also recently became aware of [mpi4py-fft](https://mpi4py-fft.readthedocs.io/en/latest/) which apparently already provides FFTW-based real-to-real transforms, although I have not tried it and am not sure what the relative benefits/drawbacks vs. pyFFTW are (aside from MPI support). There was also some interest in adding MPI support here, but I think that has stalled at the moment.

**Comment by rileyjmurray** [comment]
@grlee77 @hgomersall I have some questions for TODOs at the top of this PR.
* "Tests for builder functions and failure modes." -- I presume tests can be written in python with pyFFTW's standard framework, is that correct? 
 * "Padding and truncation for DCT / DST" -- Is this something that needs to be handled in C, or is this just a python wrapper for user input?
 * "improve ``_Xfftn``." What's that referring to?

**Comment by grlee77** [comment]
>    "Tests for builder functions and failure modes." -- I presume tests can be written in python with pyFFTW's standard framework, is that correct?

Yes.

>    "Padding and truncation for DCT / DST" -- Is this something that needs to be handled in C, or is this just a python wrapper for user input?

I think this is referring to the `n` argument to change the size of the transform ([example](https://github.com/pyFFTW/pyFFTW/pull/256/files#diff-ab165953cde05968779b909623398a3cR566-R567). Most likely this would be handled primarily in the python code for `_Xfftn` in `pyfftw/builders/_utils.py`, but I haven't looked to closely whether any change to the underlying FFTW class in Cython would also be required.

>    "improve _Xfftn." What's that referring to?

I'm not 100% sure. I copied that item from the prior PR. perhaps some of what was discussed [here](https://github.com/pyFFTW/pyFFTW/pull/95#issuecomment-207942517) and in the [older issue](https://github.com/pyFFTW/pyFFTW/issues/39).

**Comment by hgomersall** [comment]
@rileyjmurray Yes, I think the bulk of the problem is solved, and the remaining stuff can be done along the lines of what already exists in pyfftw. I think it's largely about filling in the gaps to make sure we have consistent quality across the library.

**Comment by AlbertZhangHIT** [comment]
Hi my friends,  what's the state it is rigth now? And what could I do that would help you?

**Comment by hgomersall** [comment]
Hi @AlbertZhangHIT ! The starting point is to pull @grlee77 's branch and start from there. You'll probably need to pull in from master before you begin, which hopefully will be a fairly minor job (I don't think there has been any substantial refactoring since @grlee77 's work last time).

The tasks that need doing are the unchecked boxes above. Most of the tests seem to be implemented, but I think there are some lacking for the builder functions. These might need a little bit of thought in light of the API, which I'm not really familiar with. The tests for the existing builders can be found [here](https://github.com/pyFFTW/pyFFTW/blob/master/test/test_pyfftw_builders.py) which would at least serve as a template for what is required. I suspect the [`scipy_interface`](https://github.com/pyFFTW/pyFFTW/compare/master...grlee77:r2r-try-three#diff-eddf7eb079e823fe6957bab49b5f99b9R203) tests will be very useful, since the builder API is essentially that, with some extra arguments to handle FFTW specific stuff (which is where the existing builder tests come in).

@grlee77 Would you mind clarifying some of the other points (improve _Xfftn, failure mode tests). I suspect we can merge without necessarily having all the builders in place. I think we can probably also merge without the padding and truncation bits, though I think the same strategy can probably be adopted as with normal FFT of a [wrapper object](https://github.com/pyFFTW/pyFFTW/blob/master/pyfftw/builders/_utils.py#L270).

It seems to me that n-dimensional is not yet supported and I have no opinion on that at the moment.

**Comment by hgomersall** [comment]
Do we want to change or deprecate and replace the `direction` argument to the `FFTW` class?

**Comment by grlee77** [comment]
Thanks for helping out with the summary @hgomersall 

> @grlee77 Would you mind clarifying some of the other points (improve _Xfftn, failure mode tests).

I copied those from the older #95 issue and am not sure offhand what the original intent was. Perhaps I should just delete them above.

> I suspect we can merge without necessarily having all the builders in place. I think we can probably also merge without the padding and truncation bits, though I think the same strategy can probably be adopted as with normal FFT of a wrapper object.

agreed.



@AlbertZhangHIT : If you need help with `git` commands to make a new branch from this one, preserving the commit history of past contributors, let me know.

**Comment by juhaszp95** [comment]
Hi, Thanks for the amazing package. Just a quick comment to say that I believe real-to-real transform functions of the (maybe abandoned?) mpi4py-fft package is broken, so this feature would be much appreciated here as I couldn't find any alternatives...

**Comment by hgomersall** [comment]
I think we're almost there on this... @grlee77 ?

**Comment by ashwinvis** [comment]
Hi, I came across this issue while we are planning on adding [similar functionality to fluidfft](https://foss.heptapod.net/fluiddyn/fluidfft/-/issues/31#note_153493). I saw that `scipy.fftpack` is considered legacy with the new `scipy.fft` module which relies on a pocketfft library.

- first release: https://docs.scipy.org/doc/scipy/reference/release.1.4.0.html#scipy-fft-added
- plan keyword: https://docs.scipy.org/doc/scipy/reference/release.1.5.0.html#scipy-fft-improvements
- docs: 
  - https://docs.scipy.org/doc/scipy/reference/fftpack.html#module-scipy.fftpack
  - https://docs.scipy.org/doc/scipy/reference/fft.html#module-scipy.fft

**Comment by hgomersall** [comment]
@ashwinvis I think this PR is basically ready to go. I just want a final review and sanity check before we merge.

**Comment by hgomersall** [comment]
(also, it seems, adding the changes to `scipy_fft`)

**Comment by grlee77** [comment]
I can take a quick look again and add the `scipy.fft` interface

I don't think the other remaining items above necessarily need to block merging of the PR. It is probably better to get this in sooner and we can expand lower level builder tests and support arbitrary size, `n`, for dct/dst later.

**Comment by grlee77** [comment]
This ended up being a bit more of a pain than anticipated since DCT-1 normalization for norm='ortho' was changed (fixed) in SciPy 1.2, but had not been fixed in this branch which was causing a number of test failures.

I also added support for `norm='forward'` and 'norm='backward'` that were added in SciPy 1.6.

The complex transforms should also have these new norm options, but I didn't want to add any new features not related to real-to-real transforms in this same PR.

I find the test class manipulations as done here a bit hard to follow at times and like CuPy's style used in their FFT tests. I didn't want to get into refactoring the test suite here, though!

**Comment by grlee77** [comment]
Ideally we would also move the CI that is currently on Travis-CI to Github Actions or another solution since the number of free hours on Travis are now much more limited than in the past. Let's see if it runs as is for now, though.

**Comment by hgomersall** [comment]
This is weird - why's it failing on only some python versions?

**Comment by hgomersall** [comment]
> I find the test class manipulations as done here a bit hard to follow at times and like CuPy's style used in their FFT tests. I didn't want to get into refactoring the test suite here, though!

Right, it's very much a product of it's evolution, and I think is ripe for refactoring. I think the tests are sound, but it's hard to work out what's going on.

**Comment by ashwinvis** [comment]
> This is weird - why's it failing on only some python versions?

Could be due to different scipy versions for different Python versions. Picked a few for example:

- https://travis-ci.org/github/pyFFTW/pyFFTW/jobs/757457291#L480
- https://travis-ci.org/github/pyFFTW/pyFFTW/jobs/757457292#L489
- https://travis-ci.org/github/pyFFTW/pyFFTW/jobs/757457294#L487

**Comment by hgomersall** [comment]
> > This is weird - why's it failing on only some python versions?
> 
> Could be due to different scipy versions for different Python versions. Picked a few for example:
> 
>     * https://travis-ci.org/github/pyFFTW/pyFFTW/jobs/757457291#L480
> 
>     * https://travis-ci.org/github/pyFFTW/pyFFTW/jobs/757457292#L489
> 
>     * https://travis-ci.org/github/pyFFTW/pyFFTW/jobs/757457294#L487

If that's the case, we can just feature gate to restrict it to versions that work?

**Comment by juhaszp95** [comment]
> I can take a quick look again and add the `scipy.fft` interface
> 
> I don't think the other remaining items above necessarily need to block merging of the PR. It is probably better to get this in sooner and we can expand lower level builder tests and support arbitrary size, `n`, for dct/dst later.

Agreed. I think the most important part is to be able to calculate the DCT/DST itself, both on 1-dimensional arrays, as well as on multiple dimensional arrays along (multiple) arbitrary axis, which I think should now be fine?

**Comment by ashwinvis** [comment]
The tests could be fixed too. For instance the [latest scipy (1.6.0) is available for python 3.7 to 3.9.](https://pypi.org/project/scipy/1.6.0/#files). On Travis however, there is a `$SCIPY_WHEELS` environment variable which switches between conda / pip, isn't it? This leads to inconsistencies:

- Python 3.7 was installed using conda and got scipy 1.5.2
- Python 3.8 was installed using pip and got scipy 1.6.0

**Comment by hgomersall** [comment]
I was just pondering this in #301 . I'm not familiar enough with that server to know what's on it!

**Comment by hgomersall** [comment]
I'm happy that once the tests pass on this we can merge.

**Comment by grlee77** [comment]
I need to add a version check and only test with `norm='forward'` or `norm='backward'` if SciPy >= 1.6. Let me update #285 and merge that first and then we can try again here.

**Comment by grlee77** [comment]
@ashwinvis: I updated the Travis scripts in #285 and now test with oldest supported versions for Python 3.7 and most recent versions on 3.8 and 3.9.

In local testing across SciPy 1.2, 1.4 and 1.6 I found some additional fftpack interface test failures on 1.2 due to differences in shape error handling. I included some private shape checking functions from SciPy here in 4bef86f to fix these.

**Comment by ashwinvis** [comment]
Sounds good. Let's hope it works :crossed_fingers:. I wonder:

https://github.com/pyFFTW/pyFFTW/blob/8a9e718a0b75512ab5bafa88f6084794b6f36443/setup.py#L789

Should it be 3.7 or are you relaxing the criterion since it may work in 3.6 as well?

**Comment by grlee77** [comment]
> Should it be 3.7 or are you relaxing the criterion since it may work in 3.6 as well?

I agree that it should probably be changed there as well. I guess that would prevent `pip` from trying to install from source on 3.6, though.

**Review by hgomersall — APPROVED** [review]
Great stuff. Ready to merge as far as I'm concerned.

**Comment by hgomersall** [comment]
Thanks to everyone on this rather epic 6 year old PR, particularly @drwells for the original work and @grlee77 for bringing it up to date.

**Comment by drwells** [comment]
I've long moved on to other things, but its super cool to see that this is finally done!

**Comment by juhaszp95** [comment]
Thanks everyone for making the final push - I see it's now merged into master. Would it then be possible to release this to PyPI / conda-forge so that this new feature can be easily used?

**Comment by grlee77** [comment]
A agree that we should make a new release for this. I may be able to help with it later this week and see if we can get by with the existing approach on our current free Travis credits. If we end up needing to migrate to Github Actions first that could add a little delay, but probably shouldn't be too hard to work out.

**Pull request merged**
